File: src_cpp/profile_energy.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <cmath>
#include <iomanip>
#include <random>

#include <vexcl/vexcl.hpp>
#include "fem_utils.hpp"
#include "poisson_solve.hpp"
#include "energy_kernels.hpp"

int main(int argc, char** argv) {
    std::string mesh_path = "cube_60nm_shell.npz";
    if (argc > 1) mesh_path = argv[1];

    // 1. Setup VexCL Context
    vex::Context ctx(vex::Filter::Type(CL_DEVICE_TYPE_GPU) && vex::Filter::Count(1));
    if (!ctx) {
        std::cerr << "No GPU found!" << std::endl;
        return 1;
    }
    std::cout << "Using device: " << ctx.device(0).getInfo<CL_DEVICE_NAME>() << std::endl;

    // 2. Load Mesh
    std::cout << "Loading mesh from " << mesh_path << "..." << std::endl;
    Mesh mesh = load_mesh_npz(mesh_path);
    std::cout << "Mesh Size: " << mesh.N << " nodes, " << mesh.E << " elements" << std::endl;

    // 3. Material Properties (Same as profile_energy.py)
    double Js_val = 1.0;
    double A_red = 1.0;
    double K1_red = 0.1;

    MaterialProperties props;
    int max_mat = mesh.mat_id.maxCoeff();
    props.A.assign(max_mat, 0.0);
    props.K1.assign(max_mat, 0.0);
    props.Js.assign(max_mat, 0.0);
    props.k_easy.assign(max_mat, Eigen::Vector3d(0, 0, 1));

    if (max_mat >= 1) {
        props.A[0] = A_red;
        props.K1[0] = K1_red;
        props.Js[0] = Js_val;
    }

    // 4. Assemble Matrices
    SparseMatrixCSR L, K_int, G_div, G_grad;
    assemble_matrices(mesh, props, L, K_int, G_div, G_grad);

    double vmag = compute_vmag(mesh, props);
    std::vector<double> js_v = compute_js_node_volumes(mesh, props);
    
    // 5. Setup Solver and Kernels
    std::vector<double> mask_cpu(mesh.N);
    for(int i=0; i<mesh.N; ++i) mask_cpu[i] = mesh.boundary_mask(i);
    
    PoissonSolver poisson(ctx, L, mask_cpu);
    EnergyKernels kernels(ctx, K_int, G_div, G_grad, js_v, vmag);

    // 6. Test State: Random magnetization
    std::vector<double> m_cpu(3 * mesh.N);
    std::mt19937 gen(42);
    std::normal_distribution<double> dist(0.0, 1.0);
    for (int i = 0; i < mesh.N; ++i) {
        double mx = dist(gen);
        double my = dist(gen);
        double mz = dist(gen);
        double norm = std::sqrt(mx*mx + my*my + mz*mz);
        m_cpu[3 * i + 0] = mx / norm;
        m_cpu[3 * i + 1] = my / norm;
        m_cpu[3 * i + 2] = mz / norm;
    }
    vex::vector<double> m_gpu(ctx, m_cpu);
    vex::vector<double> U_gpu(ctx, mesh.N); U_gpu = 0.0;
    vex::vector<double> b_poisson(ctx, mesh.N);
    vex::vector<double> g_gpu(ctx, 3 * mesh.N);
    Eigen::Vector3d B_ext(0.01, 0.0, 0.0);

    // Warm-up
    std::cout << "Compiling kernels (warm-up)..." << std::endl;
    kernels.compute_poisson_rhs(m_gpu, b_poisson);
    auto info_warm = poisson.solve(b_poisson, U_gpu);
    double energy_warm = kernels.energy_and_grad(m_gpu, U_gpu, B_ext, g_gpu);
    ctx.finish();
    std::cout << "Warm-up results: Energy = " << energy_warm << ", Iters = " << info_warm.first << std::endl;

    // 7. Profiling Loop 1: Full Iteration (Solve U + Kernels)
    int n_repeats = 5;
    std::cout << "\nLoop 1: Recomputing potential U every time (" << n_repeats << " iterations)..." << std::endl;
    
    double last_e = 0;
    int total_iters = 0;
    auto t0 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < n_repeats; ++i) {
        U_gpu = 0.0; // Reset initial guess to zero to match Python profile
        kernels.compute_poisson_rhs(m_gpu, b_poisson);
        auto info = poisson.solve(b_poisson, U_gpu);
        last_e = kernels.energy_and_grad(m_gpu, U_gpu, B_ext, g_gpu);
        total_iters += info.first;
    }
    ctx.finish();
    auto t1 = std::chrono::high_resolution_clock::now();
    
    std::cout << "Last Energy: " << last_e << ", Avg Poisson Iters: " << (double)total_iters/n_repeats << std::endl;
    double total_full = std::chrono::duration<double>(t1 - t0).count();
    double avg_full = total_full / n_repeats;

    // 8. Profiling Loop 2: Kernels Only (Reuse U)
    std::cout << "Loop 2: Reusing precomputed potential U (" << n_repeats << " iterations)..." << std::endl;
    
    auto t2 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < n_repeats; ++i) {
        kernels.energy_and_grad(m_gpu, U_gpu, B_ext, g_gpu);
    }
    ctx.finish();
    auto t3 = std::chrono::high_resolution_clock::now();
    
    double total_kernels = std::chrono::duration<double>(t3 - t2).count();
    double avg_kernels = total_kernels / n_repeats;

    // 9. Report
    std::cout << "\n" << std::string(40, '=') << std::endl;
    std::cout << std::left << std::setw(25) << "Metric" << " | " << std::right << std::setw(10) << "Time (ms)" << std::endl;
    std::cout << std::string(40, '-') << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << std::left << std::setw(25) << "Full Iteration (Avg)" << " | " << std::right << std::setw(10) << avg_full * 1000.0 << std::endl;
    std::cout << std::left << std::setw(25) << "Kernels Only (Avg)" << " | " << std::right << std::setw(10) << avg_kernels * 1000.0 << std::endl;
    std::cout << std::left << std::setw(25) << "Poisson Solve Overhead" << " | " << std::right << std::setw(10) << (avg_full - avg_kernels) * 1000.0 << std::endl;
    std::cout << std::string(40, '=') << std::endl;

    return 0;
}

File: benchmarking/profile_energy.py
"""profile_energy.py

Performance profiling for micromagnetic energy and gradient kernels.
Compares a full iteration (including Poisson solve) vs. energy/gradient kernels alone.
"""

from __future__ import annotations

import sys
from pathlib import Path
import time
import numpy as np
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp
from dataclasses import replace

# Add src to path for imports
sys.path.append(str(Path(__file__).parent.parent / "src"))
from fem_utils import TetGeom, compute_node_volumes
from loop import compute_volume_JinvT, compute_grad_phi_from_JinvT
from energy_kernels import make_energy_kernels
from poisson_solve import make_solve_U
import add_shell

def profile_energy():
    """Benchmark performance of energy kernels and the Poisson solver.

    Loads a 60nm cube mesh with an airbox, then measures and compares:
    1. Average time for a full iteration (Solve U + compute kernels).
    2. Average time for energy/gradient kernels alone (reusing U).
    3. Poisson solver overhead.
    """
    # 1. Load existing mesh
    mesh_path = "cube_60nm_shell.npz"
    print(f"Loading mesh from {mesh_path}...")
    data = np.load(mesh_path)
    knt, ijk = data['knt'], data['ijk']

    tets = ijk[:, :4].astype(np.int64); mat_id = ijk[:, 4].astype(np.int32)
    conn32, volume, JinvT = compute_volume_JinvT(knt, tets)
    grad_phi = compute_grad_phi_from_JinvT(JinvT)
    boundary_mask = jnp.asarray(add_shell.find_outer_boundary_mask(tets, knt.shape[0]), dtype=jnp.float64)

    geom = TetGeom(
        conn=jnp.asarray(conn32, dtype=jnp.int32),
        volume=jnp.asarray(volume, dtype=jnp.float64),
        mat_id=jnp.asarray(mat_id, dtype=jnp.int32),
        grad_phi=jnp.asarray(grad_phi, dtype=jnp.float64),
    )
    
    # 2. Material Properties
    Js = 1.0; A_red = 1.0; K1_red = 0.1
    A_lookup = jnp.array([A_red, 0.0]); K1_lookup = jnp.array([K1_red, 0.0])
    Js_lookup = jnp.array([Js, 0.0])
    k_easy = jnp.array([0.0, 0.0, 1.0]); k_easy_lookup = jnp.array([k_easy, k_easy])
    
    vol_Js = volume * np.array(Js_lookup[mat_id - 1])
    M_nodal = compute_node_volumes(replace(geom, volume=jnp.asarray(vol_Js)), chunk_elems=200_000)
    V_mag_nm = np.sum(volume[mat_id == 1])

    # 3. Kernels
    # Set tolerance to 1e-10 to match benchmark/cpp
    chunk_elems = 100_000
    solve_U = make_solve_U(geom, Js_lookup, cg_tol=1e-10, boundary_mask=boundary_mask, precond_type='amgcl', chunk_elems=chunk_elems)
    energy_and_grad, _, _ = make_energy_kernels(geom, A_lookup, K1_lookup, Js_lookup, k_easy_lookup, V_mag_nm, M_nodal, chunk_elems=chunk_elems)
    
    # Test state: Random magnetization
    key = jax.random.PRNGKey(42)
    m = jax.random.normal(key, (knt.shape[0], 3))
    m = m / jnp.linalg.norm(m, axis=1, keepdims=True)
    b_ext = jnp.array([0.01, 0.0, 0.0])

    print(f"Mesh Size: {knt.shape[0]} nodes, {tets.shape[0]} elements")
    print("Compiling kernels (warm-up)...")
    
    # Warm-up
    u_warm = solve_U(m, jnp.zeros(knt.shape[0]))
    e_warm, g_warm = energy_and_grad(m, u_warm, b_ext)
    jax.block_until_ready((u_warm, e_warm, g_warm))

    # 4. Profiling Loop 1: Full Iteration (Solve U + Kernels)
    n_repeats = 5
    print(f"\nLoop 1: Recomputing potential U every time ({n_repeats} iterations)...")
    
    t0 = time.perf_counter()
    for _ in range(n_repeats):
        u = solve_U(m, jnp.zeros(knt.shape[0]))
        e, g = energy_and_grad(m, u, b_ext)
        jax.block_until_ready((e, g))
    t1 = time.perf_counter()
    
    total_full = t1 - t0
    avg_full = total_full / n_repeats

    # 5. Profiling Loop 2: Kernels Only (Reuse U)
    print(f"Loop 2: Reusing precomputed potential U ({n_repeats} iterations)...")
    u_fixed = solve_U(m, jnp.zeros(knt.shape[0]))
    jax.block_until_ready(u_fixed)
    
    t2 = time.perf_counter()
    for _ in range(n_repeats):
        e, g = energy_and_grad(m, u_fixed, b_ext)
        jax.block_until_ready((e, g))
    t3 = time.perf_counter()
    
    total_kernels = t3 - t2
    avg_kernels = total_kernels / n_repeats

    # 6. Report
    print("\n" + "="*40)
    print(f"{'Metric':<25} | {'Time (ms)':>10}")
    print("-" * 40)
    print(f"{'Full Iteration (Avg)':<25} | {avg_full*1000:>10.2f}")
    print(f"{'Kernels Only (Avg)':<25} | {avg_kernels*1000:>10.2f}")
    print(f"{'Poisson Solve Overhead':<25} | {(avg_full - avg_kernels)*1000:>10.2f}")
    print("="*40)

if __name__ == "__main__":
    profile_energy()
